scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -619,163 +732,214 @@ class ProcessMonitor(ScalyrMonitor):
+    def record_metrics(self, pid, metrics):
+        """
+        For a process, record the metrics in a historical metrics collector
+        Collects the historical result of each metric per process in __metrics_history
+        which has form:
+
+        {
+          '<process id>: {
+                        <metric name>: [<metric at time 0>, <metric at time 1>.... ],
+                        <metric name>: [<metric at time 0>, <metric at time 1>.... ],
+                        }
+
+          '<process id>: {
+                        <metric name>: [<metric at time 0>, <metric at time 1>.... ],
+                        <metric name>: [<metric at time 0>, <metric at time 1>.... ],
+                        }
+        }
 
-        @param pid: The process id or None if there is no process to monitor.
-        @type pid: int or None
+        @param pid: Process ID
+        @param metrics: Collected metrics of the process
+        @type pid: int
+        @type metrics: dict
+        :return: None
+        """
+        for _metric, _metric_value in metrics.items():
+            if not self.metrics_history[pid].get(_metric):
+                self.metrics_history[pid][_metric] = []
+            self.metrics_history[pid][_metric].append(_metric_value)
+            # only keep the last 2 running history for any metric
+            self.metrics_history[pid][_metric] = self.metrics_history[pid][_metric][-2:]
+
+    def _reset_absolute_metrics(self):
+        """
+        At the beginning of each process metric calculation, the absolute metrics
+        need to be overwritten to the combined process(es) result. Only the cumulative
+        metrics need the previous value to calculate delta. We should set the
+        absolute metric to 0 in the beginning of this "epoch"
         """
-        if self.__pid is not None:
-            for gather in self.__gathers:
-                gather.close()
 
-        self.__pid = pid
-        self.__gathers = []
+        for pid, process_metrics in self.metrics_history.items():
+            for _metric, _metric_values in process_metrics.items():
+                if not _metric.is_cumulative:
+                    self.running_total_metrics[_metric] = 0
 
-        if pid is not None:
-            self.__gathers.append(StatReader(self.__pid, self.__id, self._logger))
-            self.__gathers.append(StatusReader(self.__pid, self.__id, self._logger))
-            self.__gathers.append(IoReader(self.__pid, self.__id, self._logger))
-            self.__gathers.append(FileDescriptorReader(self.__pid, self.__id, self._logger))
-        # TODO: Re-enable these if we can find a way to get them to truly report
-        # per-app statistics.
-        #        self.gathers.append(NetStatReader(self.pid, self.id, self._logger))
-        #        self.gathers.append(SockStatReader(self.pid, self.id, self._logger))
+    def _calculate_running_total(self, running_pids):
```

Comment:
I don't think we should call this `running_total`.  In my mind, that's something different.  I would suggest we call it `calculate_aggregated_metrics`... where it is the metrics for all for all processes.